deps/jerry/jerry-ext/handle-scope/handle-scope.c

```diff
@@ -0,0 +1,284 @@
+jerryx_escape_handle (jerryx_escapable_handle_scope scope,
+                      jerry_value_t escapee,
+                      jerry_value_t *result)
+{
+  jerryx_handle_scope parent = jerryx_handle_scope_get_parent (scope);
+  if (parent == NULL)
+  {
+    return jerryx_handle_scope_mismatch;
+  }
+
+  bool found = false;
+  do
+  {
+    size_t found_idx = 0;
+    for (size_t idx = 0; idx < JERRY_X_HANDLE_SCOPE_PRELIST_HANDLE_COUNT; idx++)
+    {
+      if (escapee == scope->handle_prelist[idx])
+      {
+        found = true;
+        found_idx = idx;
+        break;
+      }
+    }
+
+    if (found)
+    {
+      /**
+       * Escape handle to parent scope
+       * handle count has been decremented in `jerryx_hand_scope_escape_handle_from_prelist`
+       */
+      *result =jerryx_hand_scope_escape_handle_from_prelist (scope, found_idx);
+      return jerryx_handle_scope_ok;
+    }
+  }
+  while (0);
+
+  if (scope->handle_count <= JERRY_X_HANDLE_SCOPE_PRELIST_HANDLE_COUNT)
+  {
+    return jerryx_handle_scope_mismatch;
+  }
+
+  jerryx_handle_t *handle = scope->handle_ptr;
+  jerryx_handle_t *memo_handle = NULL;
+  jerryx_handle_t *found_handle = NULL;
+  while (!found)
+  {
+    if (handle == NULL)
+    {
+      return jerryx_handle_scope_mismatch;
+    }
+    if (handle->jval != escapee)
+    {
+      memo_handle = handle;
+      handle = handle->sibling;
+      continue;
+    }
+    /**
+     * Remove found handle from current scope's handle chain
+     */
+    found = true;
+    found_handle = handle;
+    if (memo_handle == NULL)
+    {
+      // found handle is the first handle in heap
+      scope->handle_ptr = found_handle->sibling;
+    }
+    else
+    {
+      memo_handle->sibling = found_handle->sibling;
+    }
+  }
+
+  /**
+   * Escape handle to parent scope
+   */
+  *result = jerryx_handle_scope_add_handle_to (found_handle, parent);
+  scope->handle_count -= 1;
+
+  return jerryx_handle_scope_ok;
+}
+
+
+/**
+ * Try to reuse given handle if possible while adding to the scope.
+ *
+ * @returns the jerry value id wrapped by given handle.
+ */
+jerry_value_t
+jerryx_handle_scope_add_handle_to (jerryx_handle_t *handle, jerryx_handle_scope scope)
+{
+  size_t handle_count = scope->handle_count;
+  scope->handle_count += 1;
+  if (handle_count < JERRY_X_HANDLE_SCOPE_PRELIST_HANDLE_COUNT)
+  {
+    jerry_value_t jval = handle->jval;
+    free (handle);
+    scope->handle_prelist[handle_count] = jval;
+    return jval;
+  }
+
+  handle->sibling = scope->handle_ptr;
+  scope->handle_ptr = handle;
+  return handle->jval;
+}
+
+
+/**
+ * Add given jerry value to the scope.
+ */
+void
+jerryx_create_handle_in_scope (jerry_value_t jval, jerryx_handle_scope scope)
+{
+  size_t handle_count = scope->handle_count;
+  scope->handle_count += 1;
+  if (handle_count < JERRY_X_HANDLE_SCOPE_PRELIST_HANDLE_COUNT)
+  {
+    scope->handle_prelist[handle_count] = jval;
+    return;
+  }
+  jerryx_handle_t *handle = malloc (sizeof(jerryx_handle_t));
```

Comment:
Do check the malloc is NULL, either.

Reply:
If `malloc` does return a NULL pointer, what error shall be thrown here?